### PR TITLE
Replace hapi validation errors with our own

### PR DESCRIPTION
**Pull request by mferno**
- [x] Write commit messages as semantic-release compatiable

### CODE REVIEW

File: src/validation-error-wrapper.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { createError } from '@springworks/error-factory';
+
+export default function(request, reply) {
+  const response = request.response;
+  if (response.isBoom && response.data.name === 'ValidationError') {
+    request.response = createError({ message: 'Validation Failed', cause: response.data });
```
Comment (mferno) [inline src/validation-error-wrapper.js]: Not entirely sure this is correct..

File: src/hapi-validation-error-replacer.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { createError } from '@springworks/error-factory';
+
+export default function(request, reply) {
+  const response = request.response;
+  if (response.isBoom && response.data.name === 'ValidationError') {
```
Comment (dbrockman) [inline src/hapi-validation-error-replacer.js]: - [ ] Did you mean `||`? Boom or Joi validation error?

Comment (mferno) [inline src/hapi-validation-error-replacer.js]: A validation is still boom right? So i'm explicitly checking that it's a validation error.

Comment (dbrockman) [inline src/hapi-validation-error-replacer.js]: Only after Hapi has passed it through `boom.wrap()`. I guess that is the case here.

Comment (dbrockman) [inline src/hapi-validation-error-replacer.js]: How about this:
```js
const error = request.response;
if (error instanceof Error) {
  request.response = createError({ message: error.message, cause: error });
}
```

**Comment by mferno** [comment]
Noticed one thing now when trying to use this in SPAPI. The hapi-basic-auth plugin will set the WWW-Auth header when authentication fails. Now we overwrite that header.

**Comment by mferno** [comment]
@sommestad when you have time. You can see how I proposed it at first in the outdated diffs. Then we thought that we might want to wrap all the errors. But that, as if you read the comment before this one, introduces a problem with the `401 Unauthorized`. Maybe only do it for the validation errors?

**Comment by sommestad** [comment]
Isn't it possible to alter the response without replacing it? Or is it possible to at least not modify the headers? Removing `WWW-Authenticate` will lead to weird things for some clients, since many depend on that header (IIRC the Android app does that, for example) instead of sending the `Authorization` header directly.

**Comment by sommestad** [comment]
For the mock API, it (probably) doesn't matter but will matter for the real ones.

**Comment by mferno** [comment]
Yeah maybe. Can have a look at that.

**Comment by dbrockman** [comment]
Use `reply` http://hapijs.com/api see onPreResponse

**Comment by mferno** [comment]
This is working now @sommestad. Though as David said, maybe the headers should be copied in the error-factory?

File: src/extension-points/pre-response.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+export default function(request, reply) {
+  const error = request.response;
+  if (error instanceof Error) {
+    const code = error.data && error.data.name === 'ValidationError' ? 422 : undefined;
+    const err = createError({ message: error.message, code: code, cause: error });
+    if (error.output && error.output.headers) {
+      err.output.headers = error.output.headers;
+    }
+    return reply(null, err);
```
Comment (mferno) [inline src/extension-points/pre-response.js]: Not sure if I should do
`return reply(null, err);`
or
`return reply(err);`

Comment (dbrockman) [inline src/extension-points/pre-response.js]: I would say `reply(err);`. It will not invoke onPreResponse again regardless

**Pull request merged**